refactor: extract UDN crawler methods and improve OOP design with dependency injection

tobybobniden opened the pull request:
## Purpose
Refactor codebase to improve OOP design patterns across crawlers and services by extracting internal methods for better testability, implementing dependency injection for services, and reorganizing test files following separation of concerns principle. Align with reference implementation patterns for improved code maintainability and extensibility.

## Changes

### UDNCrawler Enhancement (src/crawler/udn_crawler.py)
- Added `timeout` parameter to constructor (default: 10 seconds) for configurable request timeout
- Extracted `_perform_request()` private method to handle HTTP requests with error management
- Extracted `_create_search_params()` private method to construct UDN API parameters with proper URL encoding
- Extracted `_fetch_news()` private method to fetch headlines from API with data parsing
- Enhanced `parse()` method with domain validation before article parsing
- Improved error handling throughout with specific exception raising (DomainMismatchException)

### PriceCrawler Enhancement (src/crawler/price_crawler. py)
- Refactored to implement improved OOP design patterns
- Enhanced method extraction for better modularity and testability
- Improved constructor with configurable timeout parameter
- Refined error handling with specific exception management
- Better separation of concerns in crawling logic

### NewsService Dependency Injection Refactoring (src/service/news_service.py)
- Improved OOP design by implementing proper dependency injection pattern
- Enhanced constructor to accept injected crawler and database dependencies
- Decoupled service layer from direct crawler instantiation
- Improved flexibility for testing and extending service functionality
- Better code organization with clear separation of responsibilities

### Test File Refactoring - UDN Crawler
Split test_udn_crawler.py (234 lines) into two focused test files:

- test_udn_crawler_core.py (168 lines): API and parsing functionality tests (10 tests)
  - `_perform_request()` success and failure scenarios
  - `_create_search_params()` parameter construction with URL encoding validation
  - `_fetch_news()`  with single/empty results handling
  - `get_headline() ` with single and multiple pages
  - `parse()` with valid HTML, missing elements, and invalid domains
  -` _is_valid_url()` with valid and invalid domain validation

- test_udn_crawler_storage.py (56 lines): Database operation tests (3 tests)
  - `save()` with successful database operation
  - `save()` without database session
  - `save()` with database error handling

### Test File Refactoring - Price Crawler
- Organized price_crawler tests following single-responsibility principle
- Enhanced test coverage for improved OOP design
- Better test organization aligned with UDN crawler test structure

tobybobniden committed:
refactor: improve OOP design for NewsService dependency injection

- Remove redundant db parameter from NewsService methods
  * Store db as instance variable (self.db) instead of passing as parameter
  * Simplify get_article_upvote_details() and toggle_upvote() signatures

- Enhance dependency injection in router
  * Modify get_news_service() to accept and inject db Session
  * Remove unnecessary db parameter from route handlers
  * Use service.db for all database operations

- Fix test mocking issues
  * Correct mock paths from non-existent src.api.v1.news.service.OpenAI
  * Mock AIService.extract_keywords and NewsService._fetch_raw_news_data
  * Restore mock_openai helper function for test utilities

- Update .gitignore
  * Add coverage/ directory and test.db file
  * Add .coverage.* pattern for coverage data
  * Improve Python and IDE configuration ignoring

This follows OOP best practices by encapsulating db session within the service and
simplifying method signatures through proper dependency injection.

tobybobniden committed:
refactor: optimize UDN crawler implementation and split test files

- **UDNCrawler Enhancement**:
  - Add `timeout` parameter to constructor for configurable request timeout
  - Extract `_perform_request()` private method for HTTP request handling
  - Extract `_create_search_params()` private method for parameter construction
  - Extract `_fetch_news()` private method for news fetching logic
  - Add domain validation in `parse()` method before article parsing
  - Improve error handling with specific exception raising

- **Test File Refactoring**:
  - Split `test_udn_crawler.py` into two focused test files:
    - `test_udn_crawler_core.py`: API and parsing functionality (10 tests)
    - `test_udn_crawler_storage.py`: Database operations (3 tests)
  - Better separation of concerns and improved maintainability
  - Total: 13 comprehensive tests covering all UDNCrawler functionality

- Improved code maintainability and readability
- Better method granularity for easier unit testing
- Aligned with reference implementation patterns
- Test files are more focused and easier to maintain

tobybobniden committed:
fix: fix the price crawler

tobiichi3227 commented:
不是哥們，啊你怎麼把 AIService 的 refactor 也放到這個 PR 了, 然後 PR Message 還啥都沒寫

tobiichi3227 commented:
# Linus Torvalds 式 Codebase Review — price-tracker-tobybobniden

---

## 1. 核心判斷（Linus verdict）

✅ 值得做：這個 codebase 架構有潛力，但目前過度設計、分層過細，OOP 實踐不徹底，必須大幅簡化，回歸實用主義。

---

## 2. 關鍵洞察（Key Insights）

- **數據結構**：API/service/db 分層過細，service 層多數為「薄包裝」無實質邏輯，crawler OOP 實踐不徹底，資料流有重複與不必要的包裝。
- **複雜度**：service 層函數多為轉發，沒有價值；crawler 例外處理分散，容易出現 if/else 地獄。
- **風險點**：資料庫 schema 變動缺乏自動化 migration，API schemas 與前端對齊風險高，過度分層導致維護困難。

---

## 3. 致命問題（Fatal Flaws）

- service 層多數為「薄包裝」，沒有實質邏輯，純屬冗餘。
- crawler 例外處理分散，容易出現 if/else 地獄。
- db/schema 缺乏自動化 migration，未來維護風險大。
- API schemas 與前端型別對齊不嚴謹，易破壞用戶空間。

---

## 4. 改進方向（Linus-style Fixes）

1. **簡化數據結構**：service 層只保留有實質邏輯的函數，能合併就合併，router 直接調用有意義的邏輯。
2. **消除特殊情況**：crawler 例外處理集中，讓特殊情況變成正常流程，消除 if/else 地獄。
3. **自動化 migration**：db/models/schema 必須有自動化 migration，API schemas 嚴格對齊前端。
4. **函數簡潔**：所有函數不超過 3 層縮排，能用一行解決的絕不用三行。
5. **零破壞性**：保證向後兼容，任何改動都不能破壞現有 API。

---

## 5. 品味評分（Taste Score）

🟡 凑合（有潛力，但過度設計，缺乏「好品味」的簡潔）

---

## 6. Linus 式總結

這個 codebase 最大的問題不是功能不全，而是「為了分層而分層」，導致維護困難、複雜度上升。好品味的設計應該讓特殊情況消失、讓資料流自然、讓每一層都有存在的價值。現在這裡有太多「理論上的好設計」，卻沒有解決真實的痛點。

**建議：**
- 砍掉冗餘 service 層，讓 router 直接調用有意義的邏輯。
- crawler 例外處理集中，消除 if/else 地獄。
- db/schema 自動化 migration，API schemas 嚴格對齊前端。
- 所有函數不超過 3 層縮排，能用一行解決的絕不用三行。
- 任何改動都不能破壞現有 API，永遠不破壞用戶空間。

---

【這不是在解決不存在的問題。真正的問題是：複雜度過高、維護困難、資料流不清。】

---

Poloqwq commented:
# @#$&式 Code review -- price-tracker-tobybobniden
---
## 心得感想
為什麼是這樣呢？經過上述討論，羊排鮭魚與波羅游泳，到底應該如何實現。莎士比亞曾經告訴世人，意志命運往往背道而馳，決心到最後會全部推倒。這句話語雖然很短，但令我浮想聯翩。林則徐有講過一句名言，豈能盡如人意，但求無愧我心！這似乎非常的有道理，對吧？羊排鮭魚與波羅游泳因何而發生？更多羊排鮭魚與波羅游泳的意義是這樣的，我認為，羊排鮭魚與波羅游泳的意義其實就隱藏在我們的生活中，我們都知道，只要有意義，那麽就必須慎重考慮。要想清楚，羊排鮭魚與波羅游泳，到底是一種怎麽樣的存在。西班牙人有說過，自己的鞋子，自己知道緊在哪里。帶著這句話，我們還要更加慎重的審視這個問題：那麽，老子曾經說過，知人者智，自知者明。勝人者有力，自勝者強。這不禁令我深思。我這幾天一直在思索這個問題，伏契克有一句名言，為了爭取將來的美好而犧牲了的人，都是一尊石質的雕像。這讓我深深地想到，其實羊排鮭魚與波羅游泳是非常值得我們深思的。所以說，了解清楚羊排鮭魚與波羅游泳到底是一種怎麽樣的存在，是解決一切問題的關鍵。至於為什麼要思考羊排鮭魚與波羅游泳呢？其實是有更深層的原因，而這些並不是完全重要，更加重要的問題是，屠格涅夫有一句名言，凡事只要看得淡些，就沒有什麼可憂慮的了，只要不因憤怒而誇大事態，就沒有什麼事情值得生氣的了。這讓我深深地想到，羊排鮭魚與波羅游泳，發生了會如何，不發生又會如何。每個人都不得不面對這些問題。 在面對這種問題時，就我個人來說，羊排鮭魚與波羅游泳對我的意義，不能不說非常重大。羊排鮭魚與波羅游泳的發生，到底需要如何做到，不羊排鮭魚與波羅游泳的發生，又會如何產生。這種事實對本人來說意義重大，相信對這個世界也是有一定意義的。你真的了解羊排鮭魚與波羅游泳嗎？羊排鮭魚與波羅游泳真的是很值得探究，一般來說，既然如此，我們一般認為，抓住了問題的關鍵，其他一切則會迎刃而解。我們不得不面對一個非常尷尬的事實，那就是，歌德有講過一句名言，讀一本好書，就如同和一個高尚的人在交談。這讓我深深地想到，一般來講，我們都必須務必慎重的考慮考慮。現在，解決羊排鮭魚與波羅游泳的問題，是非常非常重要的。 所以，在這種困難的抉擇下，本人思來想去，寢食難安。生活中，若羊排鮭魚與波羅游泳出現了，我們就不得不考慮它出現了的事實。既然是這樣，所謂羊排鮭魚與波羅游泳，關鍵是羊排鮭魚與波羅游泳需要如何寫。總結的來說，本人也是經過了深思熟慮，在每個日日夜夜思考這個問題。克勞斯·莫瑟爵士在不經意間這樣說過，教育需要花費錢，而無知也是一樣。我希望諸位也能好好地體會這句話。問題的關鍵究竟為何？為什麼羊排鮭魚與波羅游泳對我們來說這麼重要？
## 總結
總評分級

本次 PR 評級：Fuiyooooo

> 設計思路是對的，拆分、注入、測試方向都很健康，
> 只是 scope creep 跟 session lifecycle 還差一點最後收尾。
> 修完 checklist 之後，很有機會往「牛逼666電神鹽都不鹽」邁進。

評級說明（由上到下）：

1. **牛逼666電神鹽都不鹽**
—— 看到會懷疑自己是不是該轉行，完全挑不到像樣的毛病，
只好硬擠一點「也許未來可以考慮…」這種禮貌性建議。

2. **Fuiyooooo**
—— 整體方向正、架構穩、有明顯用心，
有幾個技術債與風險點需要處理，但屬於「修一修就很好 merge」的級別。

3. **No flavor haiyaaaa**
—— 功能正常、邏輯也說得過去，但整體像白開水：
沒大問題，也沒什麼記憶點，看得出來作者當時主要想趕進度。

4. **Brotha Uuhh, What's this brotha D:**
—— 功能可能還勉強跑得動，但架構、命名、錯誤處理都在求救。
這種等級的 PR 通常會搭配一句「我們可以約個時間一起看一下嗎？」

5. **your cousin Ling Ling invented Calculus 6 years old, and what are you? FAILUREEEE**
—— 會讓 reviewer 陷入人生反思的等級。
這種 PR 的建議修正方式通常不是「改一改」，
而是「先關掉，另外開一個 branch 重寫會比較快」。

sgtsai commented:
B. XOR Array
time limit per test1 second
memory limit per test256 megabytes

You are given three integers n
, l
, and r
.

You need to generate an array a
 of positive (1≤ai≤109
) integers of length n
. Let f(x,y)
, for 1≤x≤y≤n
, be the bitwise XOR∗
 value ax⊕ax+1⊕…⊕ay
. You need to make sure that
{f(x,y)=0when x=l and y=r;andf(x,y)≠0when x≠l or y≠r.

∗
⊕
 denotes the [bitwise XOR operation](https://en.wikipedia.org/wiki/Bitwise_operation#XOR).

Input
Each test contains multiple test cases. The first line contains the number of test cases t
 (1≤t≤104
). The description of the test cases follows.

The only line of each test case contains three integers n
, l
, and r
 (2≤n≤4⋅105
, 1≤l<r≤n
).

It is guaranteed that the sum of n
 across all test cases does not exceed 5⋅105
.

Output
For each test case, print a single line containing n
 integers a1,a2,…an
.

We can show that an answer always exists. If there are multiple solutions, print any of them.

Example
InputCopy
4
3 1 3
4 1 3
8 2 4
4 3 4
OutputCopy
9 8 1 
2 7 5 4 
9 1 9 8 10 5 4 9
85484 130377 6031 6031
Note
In the first test case, f(1,3)=9⊕8⊕1=0
, while all other non-empty subarrays have non-zero bitwise XOR:

f(1,2)=9⊕8=1≠0
,
f(2,3)=8⊕1=9≠0
,
f(1,1)=9≠0
,
f(2,2)=8≠0
,
f(3,3)=1≠0
.
In the second test case, 2⊕7⊕5=0
, while, for example, 7⊕5⊕4=6≠0
.



[Codeforces](https://codeforces.com/) (c) Copyright 2010-2025 Mike Mirzayanov
The only programming contests Web 2.0 platform
Server time: Dec/11/2025 08:14:09UTC+8 (i1).
Desktop version, switch to [mobile version](https://codeforces.com/problemset/problem/2175/B?mobile=true).
[Privacy Policy](https://codeforces.com/privacy) | [Terms and Conditions](https://codeforces.com/terms)
Supported by

iceice666 commented:
# 🔥 Linus Review Mode

**Immediate Verdict:** NAK. What the hell is this thread-safety nightmare?

**Technical Breakdown:**

1.  **Concurrency Suicide in `router.py`**:
    You have a global singleton `news_service` initialized in `main.py`. Then, in `src/api/v1/news/router.py`, you have this monstrosity:
    ```python
    def get_news_service(db: Session = Depends(get_db)) -> NewsService:
        """依賴注入 NewsService 實例，注入 db session"""
        from src.main import news_service
        news_service.db = db  # <--- HERE. RIGHT HERE.
        return news_service
    ```
    Are you out of your mind? You are mutating the state of a **GLOBAL SINGLETON** inside a per-request dependency. In a multi-threaded environment (which any real web server is), Request A comes in, sets `news_service.db`, then Request B comes in and overwrites `news_service.db` while Request A is still using it. Request A tries to commit and explodes, or worse, commits to the wrong transaction context. This is fundamentally broken. It's not "dependency injection"; it's "race condition injection."

2.  **Crawlers Saving Data? Seriously?**:
    In `src/crawler/base.py` and `udn_crawler.py`:
    ```python
    @staticmethod
    def save(news: News, db: Session = None) -> bool:
    ```
    Why the f*ck does a `Crawler` class know about SQLAlchemy sessions? A crawler's job is to fetch URL bytes and parse them into data structures. **Period.** It should not know what a database is. It should return the `News` object, and a separate Service or Repository layer handles the storage. You've violated the Single Responsibility Principle so hard it's crying for its mother. And making it a `staticmethod`? This is just procedural garbage wrapped in a class for no reason.

3.  **Transaction Boundary Violation**:
    Your `UDNCrawler.save` method calls `db.commit()`.
    ```python
    db.add(article)
    db.commit()
    ```
    If I call this method inside a larger transaction context in the Service layer, you've just forced a commit prematurely. You don't control the transaction in a utility method! You bubble the exception up or let the caller handle the unit of work.

4.  **Java-itis in `PriceService`**:
    `PriceService` is just a collection of `@staticmethod`s.
    ```python
    class PriceService:
        @staticmethod
        def get_necessities_prices(...)
    ```
    This is Python, not Java. If a class has no state and only static methods, it's called a **MODULE**. Stop wrapping functions in classes just to look "OOP." It's just noise.

**Consequences:**
*   **Data Corruption:** The race condition in `router.py` guarantees that under load, users will get 500 errors, or worse, cross-talk between database sessions.
*   **Untestable Spaghettti:** By coupling the Crawler to the Database, you can't test the scraping logic without mocking an entire DB session.
*   **Unmaintainable Mess:** The "Dependency Injection" here is fake. You're just monkey-patching a singleton at runtime.

**Dismissal:**
**HELL NO.**
1.  **Fix the Service scope.** Either `NewsService` is transient (created per request) so you can safely pass the `db` session to its constructor, OR `NewsService` is a singleton and you pass the `db` session to the **methods** that need it, never storing it on `self`.
2.  **Strip `save()` from the Crawler.** The crawler returns data. The Service saves data.
3.  **Delete `PriceService` class.** Make them functions in `src/api/v1/prices/service.py`.

Do not merge this until you understand how variable scope works.

Cecill65535 commented:
B. Two Sets
time limit per test1 second
memory limit per test256 megabytes

Little X has n distinct integers: p1, p2, ..., pn. He wants to divide all of them into two sets A and B. The following two conditions must be satisfied:

If number x belongs to set A, then number a - x must also belong to set A.
If number x belongs to set B, then number b - x must also belong to set B.
Help Little X divide the numbers into two sets or determine that it's impossible.

Input
The first line contains three space-separated integers n, a, b (1 ≤ n ≤ 105; 1 ≤ a, b ≤ 109). The next line contains n space-separated distinct integers p1, p2, ..., pn (1 ≤ pi ≤ 109).

Output
If there is a way to divide the numbers into two sets, then print "YES" in the first line. Then print n integers: b1, b2, ..., bn (bi equals either 0, or 1), describing the division. If bi equals to 0, then pi belongs to set A, otherwise it belongs to set B.

If it's impossible, print "NO" (without the quotes).

Examples
input
4 5 9
2 3 4 5
output
YES
0 0 1 1
input
3 3 4
1 2 4
output
NO

Note
It's OK if all the numbers are in the same set, and the other one is empty.

tobybobniden committed:
Refactor backend services, crawlers, and fix API schemas

- **Service Layer**: Refactored `NewsService` to be stateless (dependency injection for DB session) and integrated `UDNCrawler` directly. Refactored `PriceService` to use Pydantic models and proper HTTP exception handling.
- **Crawler**: Centralized exception handling in `UDNCrawler` using a new decorator in `src/crawler/utils.py`.
- **Database**: Fixed `alembic/env.py` to correctly detect modelRefactor backend services, crawlers, and fix API schemas

- **Service Layer**: Refactored `NewsService` to be stateless (dependency injection for DB session) and integrated `UDNCrawler` directly. Refactored `PriceService` to use Pydantic models and proper HTTP exception handling.
- **Crawler**: Centralized exception handling in `UDNCrawler` using a new decorator in `src/crawler/utils.py`.
- **Database**: Fixed `alembic/env.py` to correctly detect model

tobybobniden committed:
fix: fix all CI poops
